heat/core/operations.py

```diff
@@ -544,7 +549,620 @@ def sum(x, axis=None, out=None):
+        mu_tot_hold = tensor.zeros(target_dimens)
+        mu_tot = tensor.zeros(target_dimens)
+
+        beg = 0
+        if isinstance(dimen, (list, tuple, np.ndarray, tensor.tensor)):
+            # multiple dimensions
+            score = [True for itt in dimen if itt < x.split].count(True)
+        elif isinstance(dimen, (int, float)):
+            # one dimension
+            score = 1 if dimen < x.split else 0
+        for r in range(x.comm.Get_size()):
+            # need to add them together in order
+            if x.comm.rank == r:
+                end = beg + x.lshape[x.split]
+                if x.split - score == 0:
+                    mu_tot_hold[beg:end] = mu_proc
+                elif x.split - score == 1:
+                    mu_tot_hold[:, beg:end] = mu_proc
+                elif x.split - score == 2:
+                    mu_tot_hold[:, :, beg:end] = mu_proc
+                elif x.split - score == 3:
+                    mu_tot_hold[:, :, :, beg:end] = mu_proc
+                elif x.split - score == 4:
+                    mu_tot_hold[:, :, :, :, beg:end] = mu_proc
+                else:
+                    print('else', mu_tot_hold[beg:end].lshape, mu_proc.lshape)
+                    mu_tot_hold[beg:end] = mu_proc
+                beg = end
+            beg = x.comm.bcast(beg, r)
+        x.comm.Allreduce(mu_tot_hold, mu_tot, MPI.SUM)
+        x.comm.barrier()
+        return mu_tot
+
+    def reduce_means_elementwise(target_dimens):
+        """
+        Function to combine the calculated means together.
+        This does an element-wise update of the calculated means to merge them together using the merge_means function.
+        This function operates using x from the mean function paramters
+
+        Parameters
+        ----------
+        target_dimens : iterable
+                        iterable with the dimensions of the output of the mean function
+
+        Returns
+        -------
+        ht.tensor of the calculated means
+        """
+
+        mu = __local_operation(torch.mean, x, out=None, **{'dim': dimen})
+
+        mu_to_combine = tensor.zeros((x.comm.Get_size(),) + tuple(target_dimens))
+        mu_tot = tensor.zeros((x.comm.Get_size(),) + tuple(target_dimens))
+
+        n_for_merge = tensor.zeros(x.comm.Get_size())
+        n2 = tensor.zeros(x.comm.Get_size())
+        mu_to_combine[x.comm.Get_rank()] = mu
+        n2[x.comm.Get_rank()] = x.lshape[x.split]
+        x.comm.Allreduce(mu_to_combine, mu_tot, MPI.SUM)
+        x.comm.Allreduce(n2, n_for_merge, MPI.SUM)
+
+        sz = x.comm.Get_size()
+        rem1 = rem2 = 0
+        while True:
+            if sz % 2 != 0:  # test if the length is an odd number
+                if rem1 != 0 and rem2 == 0:
+                    rem2 = sz - 1
+                elif rem1 == 0:
+                    rem1 = sz - 1
+            splt = int(sz / 2)
+            for sp_it in range(splt):  # todo: multithread perfect opp for GPU parrallelizm
```

Comment:
issue as ticket or fix in PR

Reply:
there is a GPU course at JSC coming up and I plan to use this as my problem to be solved